Applications: add migration guide/tutorials

stephan-merkli opened the pull request:
Add detailed description of how to migrate from the old module structure (.war) to the new one (.app.zip/.app.image) including replacement of web.xml. Add tutorial on how to deploy with shell scripts as service and by using docker images.

369376

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+
+include::common:partial$_header.adoc[]
+
+A different approach then to use the output of the `.app.zip` module builds is to build docker images instead.
```
```suggestion
A different approach than to use the output of the `.app.zip` module builds is to build docker images instead.
```

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+
+A different approach then to use the output of the `.app.zip` module builds is to build docker images instead.
+The docker images might be used within a Kubernetes setup or similar to run your application.
+In this guide we use a local Docker installation instead.
```
```suggestion
In this guide we use a local Docker installation.
```

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+== Build and push Docker images
+
+For the `.app.image` modules to be built, the Maven profile `exec.docker.image` must be activated.
+This manual activation is to avoid that a full Maven build of the project would fail because usually there is no local Docker registry running that could accept the built images.
```
```suggestion
This manual activation is to avoid that a full Maven build of the project would fail because usually there is no local Docker registry running that would accept the built images.
```

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+
+  mvn clean install -Pexec.docker.image -Djib.allowInsecureRegistries=true
+
+We skip tests for this Maven build (`-DskipTests`),
```
Shouldn't `-DskipTests` be also specified above (line 38)?

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+activate the required Maven profile (`-Pexec.docker.image`)
+and tell Google JIB to allow insecure registries (`-Djib.allowInsecureRegistries=true`) because our local registry is only running on `http`.
+
+Check the output of the built, it should say something as
```
```suggestion
Check the output of the build, it should say something as
```

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+
+== Run the servers
+
+Because the frontend server needs to be able to communication with the backend server, we create an own network
```
```suggestion
Because the frontend server needs to be able to communicate with the backend server, we create an own network
```

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+== Setup Local Registry
+
+The Maven property `docker.app.image.registry` (in the parent `pom.xml`) defines where the docker image is pushed to.
+By default, `localhost:5000` is used which represent a locally running Docker registry without authentication.
```
```suggestion
By default, `localhost:5000` is used which represents a locally running Docker registry without authentication.
```

matthiaso commented on docs/modules/getstarted/pages/run-as-docker-container.adoc:
```diff
@@ -0,0 +1,108 @@
+
+In the `config.properties` of your frontend server (`helloscout.ui.html.app.image`) the property `scout.backendUrl` is not defined (because not known before).
+Within a docker network, the containers are accessible by their names (`--name` argument when container is run).
+To set/override property values via environment variables, use the `-e` argument (i.e. `-e scout.backendUrl=http://helloscout-server:8080`, environment variables use `_` instead of `.`).
```
```suggestion
To set/override property values via environment variables, use the `-e` argument (i.e. `-e scout_backendUrl=http://helloscout-server:8080`, environment variables use `_` instead of `.`).
```

Shouldn't underscore (_) be used here already as the -e switch suggests an environment variable?

matthiaso commented on docs/modules/releasenotes/partials/release-notes-24.1.adoc:
```diff
@@ -118,4 +118,10 @@ During form-based login, for second-factor enabled accounts, a token is requeste
+The former `org.eclipse.scout.dev.jetty.JettyServer` (now `Application`) was so far only used for development environments.
+With this release it was improved so that it can be used in a productive environment too.
+The Scout Application class (`org.eclipse.scout.rt.app.Application`) contains a main-method which may be started directly and itself starts an embedded Jetty server.
+Therefore, no more .war files need to be built that runs in an application server.
```
```suggestion
Therefore, no more .war files need to be built that run in an application server.
```

matthiaso reviewed (commented):
Intermediate

matthiaso commented on docs/modules/getstarted/pages/run-as-service.adoc:
```diff
@@ -0,0 +1,281 @@
+This is why the `Hello Scout` applications are configured to expect HTTPS between the browser and the frontend server for production by default.
+
+In case you don't want to use HTTPS the application must be slightly modified.
+Then you can skip the next two section and instead follow the instructions in section <<run-as-service-work-with-http>>.
```
```suggestion
Then you can skip the next two sections and instead follow the instructions in section <<run-as-service-work-with-http>>.
```

matthiaso commented on docs/modules/getstarted/pages/run-as-service.adoc:
```diff
@@ -0,0 +1,281 @@
+== Frontend-backend server communication over an encrypted channel
+
+The communication between the frontend and the backend server isn't encrypted by default
+because they might be running on the same machine or in a protected environment otherwise and no access to the backend server is available from "outside".
```
```suggestion
because they might be running on the same machine or in a protected environment and no access to the backend server is available from "outside".
```

matthiaso commented on docs/modules/getstarted/pages/run-as-service.adoc:
```diff
@@ -0,0 +1,281 @@
+
+  keytool -exportcert -alias helloscout_backend_localhost -storepass changeit_store -keystore /opt/helloscout/certs/helloscout_backend_localhost.jks -file /opt/helloscout/certs/helloscout_backend_localhost.der
+
+In the last step add the self-signed certificate to the known certificates of the Java runtime.
+Make sure you modify the `cacerts` file of the Java runtime that is used to run your application and modify the path to the `cacerts` file accordingly.
+
+  keytool -import -alias helloscout_backend_localhost -trustcacerts -storepass changeit_store -keystore C:\java\jdk11\lib\security\cacerts -file /opt/helloscout/certs/helloscout_backend_localhost.der
```
We could as well use `org.eclipse.scout.rt.server.commons.ServerCommonsConfigProperties.TrustedCertificatesProperty` and configure the trusted certificate just for our application? But adding it to `cacerts` should work as well.

stephan-merkli replied:
Thanks, I use this property instead, easier to configure and should be sufficient, no need to register within the JVM.

matthiaso commented on docs/modules/getstarted/pages/run-as-service.adoc:
```diff
@@ -0,0 +1,281 @@
+** Add the property `scout.app.sessionCookieConfigSecure=false` to remove the secure flag on the session cookie.
+** Add the property `scout.auth.cookieSessionValidateSecure=false` to disable the check for an encrypted channel (HTTPS).
+
+More on this topic can be found in the Scout Architecture Documentation.
```
Maybe add a hyperlink where more information could be found?

stephan-merkli replied:
Removed the sentence, was there before, not sure where more information can be found.

matthiaso commented on docs/modules/migration/partials/migration-guide-24.1.adoc:
```diff
@@ -231,6 +233,322 @@ The following renaming is applied.
+}
+----
+
+Applies to other filters too (init-params will be different/not exist).
```
```suggestion
Applies to other filters too (init-params may be different/may not exist).
```

matthiaso commented on docs/modules/migration/partials/migration-guide-24.1.adoc:
```diff
@@ -231,6 +233,322 @@ The following renaming is applied.
+}
+----
+
+Applies to other servlets too (init-params will be different/not exist, load-on-startup flag too).
```
```suggestion
Applies to other servlets too (init-params may be different/may not exist, load-on-startup flag too).
```

matthiaso commented on docs/modules/releasenotes/partials/release-notes-24.1.adoc:
```diff
@@ -90,3 +90,11 @@ To get familiar with Scouts new accessibility features we recommend using `aria.
+The former `org.eclipse.scout.dev.jetty.JettyServer` (now `Application`) was so far only used for development environments.
+With this release it was improved so that it can be used in a productive environment too.
+The Scout Application class (`org.eclipse.scout.rt.app.Application`) contains a main-method which may be started directly and itself starts an embedded Jetty server.
+Therefore, no more .war files need to be built that run in an application server.
```
Even better: *no more additional application server is required*

stephan-merkli replied:
Replaced by "Therefore, an application server isn't required anymore."

matthiaso reviewed (approved):
My testing of `docs/modules/getstarted/pages/run-as-docker-container.adoc` was successful

stephan-merkli committed:
Applications: add migration guide/tutorials

Add detailed description of how to migrate from the old module structure (.war) to the new one (.app.zip/.app.image) including replacement of web.xml.
Add tutorial on how to deploy with shell scripts as service and by using docker images.

369376
